genume/registry/child.py

```diff
@@ -0,0 +1,88 @@
+from shlex import split as cmdsplit
+import subprocess as sbpr
+import logging as log
+import time
+import os
+
+from genume.constants import VERSION, NAME, DESC, SCRIPTS_BASH_EXTRA
+from genume.utils import merge_dicts, find_key_of_value
+from genume.registry.category import CategoryEntry
+from genume.registry.parser import COMMAND_REGISTRY
+
+
+class ChildHandler:
+    """Handles child lifecycle and command parsing."""
+
+    def __init__(self, path, root_cat, parent):
+        self.path = path
+        self.root = root_cat
+        self.shadow = CategoryEntry()
+        self.state = {}
+        self.observer = parent.observer
+        self.done = False
+
+    def generate_env(self):
+        "Generates the environment variables that are passed to the child"
+        name = "root"
+        if self.root.parent is not None:
+            name = find_key_of_value(root.parent, root)
+        extra_env = {"HOST_VERSION": VERSION, "HOST_NAME": NAME,
+                     "HOST_DESC": DESC, "MASTER_CATEGORY": name}
+        env = merge_dicts(extra_env, os.environ)
+        # Generate new path.
+        env["PATH"] = SCRIPTS_BASH_EXTRA + ":" + env["PATH"]
+        return env
+
+    def start(self):
+        "Prepares and launches the child."
+        child = sbpr.Popen(args=str(self.path), shell=False, bufsize=0, universal_newlines=True,
+                           stdin=sbpr.PIPE, stdout=sbpr.PIPE, close_fds=False, env=self.generate_env())
+        log.debug("Started child \"%s\" with pid %i" %
+                  (self.path.name, child.pid))
+        self.executable = child
+        self.start = time.time_ns()
+        # Return self for chaining support.
+        return self
+
+    def check_status(self):
+        "Returns true if the executable has finished execution."
+        if self.executable.poll() is not None:
+            delta_ms = (time.time_ns() - self.start) / 1000000
+            if self.executable.returncode == 0:
+                log.info("Child %s (%i) has been executed successfully in %.2f ms!" % (
+                    self.path.name, self.executable.pid, delta_ms))
+            else:
+                log.error("Child %s (%i) has stopped execution with code %i." % (
+                    self.path.name, self.executable.pid, self.executable.returncode, delta_ms))
+            self.done = True
+            return True
+        else:
+            return False
+
+    def do_step(self):
+        "Performs one command proccessing step. Returns true if the command finished execution else false."
+        # Trying to avoid any potential deadlocks...
+        if self.check_status():
+            return True
```

Comment:
This is a bug, the script may have exited before all the stdout have been read.

You would suggest something like:
```diff
-        # Trying to avoid any potential deadlocks...
-        if self.check_status():
-            return True
         # Get next command.
         command_text = self.executable.stdout.readline()
+
+        # Trying to avoid any potential deadlocks...
+        if command_text == '' and self.check_status():
+            return True
+
         command_stack = cmdsplit(command_text)
```
and then at the default return:
```diff
-        return self.check_status()
+        return False
```

Reply:
Ok